### PR TITLE
Redshift Proof System Implementation

### LINKED ISSUE #20: Implement Redshift Proof System

### PR DESCRIPTION
**Pull request by nkaskov**
Resolves #20 and https://github.com/NilFoundation/evm-mina-verification/issues/1 along with https://github.com/NilFoundation/evm-solana-verification/issues/1

### CODE REVIEW

File: include/nil/crypto3/zk/snark/commitments/batched_kate.hpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                 * Oecumenical Noninteractive arguments of Knowledge",
+                 * Ariel Gabizon, Zachary J. Williamson, Oana Ciobotaru,
+                 * Aztec,
+                 * <https://eprint.iacr.org/2019/953.pdf>
+                 */
+                class batched_kate_commitment_scheme : commitment_scheme<...> {
+                    typedef TCommitment...;
+                    typedef TDecommitmentInfo...;
+                    typedef TSRS...;
+                    typedef TData...;
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/commitments/batched_kate.hpp:51]: Anything is supposed to be here?

File: include/nil/crypto3/zk/snark/commitments/batched_kate.hpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+                public:
+                    virtual std::pair<TCommitment, TDecommitmentInfo> commit(TSRS PK, TData phi) {
+                    }
+
+                    virtual... open(TSRS PK, TCommitment C, TData phi, TDecommitmentInfo d) {
+                    }
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/commitments/batched_kate.hpp:58]: Anything is supposed to be here?

File: include/nil/crypto3/zk/snark/commitments/fri_commitment.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                 * Assimakis Kattis, Konstantin Panarin, Alexander Vlasov,
+                 * Matter Labs,
+                 * <https://eprint.iacr.org/2019/1400.pdf>
+                 */
+                template<std::size_t d, std::size_t Rounds>
+                class fri_commitment_scheme : commitment_scheme<...> {
+                    typedef std::array<math::polynomial<...>, Rounds> TCommitment;
+                    typedef std::array<std::size_t, d + 1> TDecommitmentInfo;
+                    typedef... TSRS;
+                    typedef std::tuple<math::polynomial<...>, std::array<std::size_t, Rounds>>
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/commitments/fri_commitment.hpp:54]: Anything is supposed to be here?

File: include/nil/crypto3/zk/snark/systems/plonk/kate/proof.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+namespace nil {
+    namespace crypto3 {
+        namespace zk {
+            namespace snark {
+
+                template<typename TCurve, typename TCommitment>
+                class plonk_proof;
+
+                template<typename TCurve>
+                struct plonk_proof<TCurve, batched_kate_commitment_scheme<...>> {
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/kate/proof.hpp:40]: Should we finish this?

File: include/nil/crypto3/zk/snark/systems/plonk/kate/proving_key.hpp [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+                        math::polynomial w_2_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+                        math::polynomial w_3_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+                        math::polynomial w_4_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+                        math::polynomial z_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+
+                        memset((void *)&w_1_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
+                        memset((void *)&w_2_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
+                        memset((void *)&w_3_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
+                        memset((void *)&w_4_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
+                        memset((void *)&z_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/kate/proving_key.hpp:206]: Can we really do this in here?

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/generator.hpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                    //     3> &k){
+                    //     ...
+                    // }
+
+                    // static inline math::polynomial::polynom<...> sigma_p2_permutation(
+                    //     math::polynomial::polynom<...> input, std::size_t n, std::array<typename
+                    //     TCurve::scalar_field_type::value_type, 3> &k){
+
+                    //     return (tau(sigma_p1_permutation(tau_reverted(input, n, k), n, k), n, k));
+                    // }
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/redshift/generator.hpp:92]: Should we do anything with this?

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+                        std::size_t N_wires = ...;
+                        std::size_t N_perm = ...;
+                        std::size_t N_sel = ...;
+                        std::size_t N_const = ...;
+
+                        fiat_shamir_heuristic<transcript_manifest, hashes::sha2> transcript;
+
+                        ... setup_values = ...;
+                        transcript(setup_values);
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp:70]: This seems to be not defined by the spec.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                        typename transcript_hash_type::digest_type upsilon_bytes =
+                            transcript.get_challenge<transcript_manifest::challenges_ids::upsilon>();
+
+                        typename FieldType::value_type upsilon = algebra::marshalling<FieldType>(upsilon_bytes);
+
+                        std::array<typename FieldType::value_type, k> fT_evaluation_points = {upsilon};
+
+                        for (std::size_t i = 0; i < N_wires; i++) {
+                            if (!lpc::verify_eval(fT_evaluation_points, proof.f_commitments[i], proof.f_lpc_proofs[i],
+                                                  ...)) {
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp:108]: Should we do something with this?

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+                                return false;
+                            }
+                        }
+
+                        std::array<typename FieldType::value_type, k> PQ_evaluation_points = {upsilon, upsilon * omega};
+                        if (!lpc::verify_eval(PQ_evaluation_points, proof.P_commitment, proof.P_lpc_proof, ...)) {
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp:114]: Same in here.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+                            return false;
+                        }
+                        if (!lpc::verify_eval(PQ_evaluation_points, proof.Q_commitment, proof.Q_lpc_proof, ...)) {
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp:117]: Same in here.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+                            return false;
+                        }
+
+                        for (std::size_t i = 0; i < N_perm + 1; i++) {
+                            if (!lpc::verify_eval(fT_evaluation_points, proof.T_commitments[i], proof.T_lpc_proofs[i],
+                                                  ...)) {
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp:123]: Same in here.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+                        std::array<math::polynomial::polynom<typename FieldType::value_type>, 6> F;
+                        F[0] = verification_key.L_basis[1] * (P - 1);
+                        F[1] = verification_key.L_basis[1] * (Q - 1);
+                        F[2] = P * p_1 - (P << 1);
+                        F[3] = Q * q_1 - (Q << 1);
+                        F[4] = verification_key.L_basis[n] * ((P << 1) - (Q << 1));
+                        F[5] = verification_key.PI;
+
+                        for (std::size_t i = 0; i < N_sel; i++) {
+                            F[5] += q[i] * ....gate[i];
```
Comment (nemothenoone) [inline include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp:137]: Same in here.

**Review by nemothenoone — CHANGES_REQUESTED** [review]

File: include/nil/crypto3/zk/snark/systems/plonk/kate/proving_key.hpp (hunk 5 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/kate/proving_key.hpp:206]: Kate PLONK system seems to be not required at the moment. We decided to leave it in this state for some time.

File: include/nil/crypto3/zk/snark/commitments/batched_kate.hpp (hunk 1 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/commitments/batched_kate.hpp]: Fixed.

File: include/nil/crypto3/zk/snark/commitments/fri_commitment.hpp (hunk 3 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/commitments/fri_commitment.hpp]: Fixed.

File: include/nil/crypto3/zk/snark/systems/plonk/kate/proof.hpp (hunk 4 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/kate/proof.hpp:40]: Fixed.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/generator.hpp (hunk 6 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/redshift/generator.hpp]: Fixed.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp (hunk 7 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp]: Fixed.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp (hunk 8 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp]: Fixed.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp (hunk 9 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp]: Fixed.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp (hunk 12 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp]: Fixed.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp (hunk 11 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp]: Fixed.

File: include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp (hunk 10 above)
Comment (nkaskov) [reply include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp]: Fixed.

**Review by nemothenoone — APPROVED** [review]

**Pull request merged**